esp-hal/src/radio_clock_ctrl.rs

```diff
@@ -0,0 +1,171 @@
+        critical_section::with(|cs| {
+            let mut radio_clock = self.radio_clock.borrow_ref_mut(cs);
+            if enabled {
+                radio_clock.enable(radio_peripheral)
+            } else {
+                radio_clock.disable(radio_peripheral)
+            }
+        })
+    }
+    /// Enable or disable the PHY clock.
+    ///
+    /// If the PHY clock state was successfully changed, or the current state
+    /// already matched the specified one `Ok(())` is returned.
+    /// If the PHY clock is being disabled, but other modem clocks are still
+    /// active `Err(ModemClocksEnabledError)` is returned.
+    fn set_phy_clock_status(&self, enabled: bool) -> Result<(), ModemClocksEnabledError> {
+        // We shouldn't disable the PHY clock, if other clocks are still active.
+        if !enabled && self.any_modem_clocks_enabled() {
+            return Err(ModemClocksEnabledError);
+        }
+        // If the current clock status is already the same as the user provided one, we
+        // return. Otherwise we store the new value.
+        if PHY_CLOCK_ENABLED
+            .compare_exchange(!enabled, enabled, Ordering::Relaxed, Ordering::Relaxed)
+            .is_err()
+        {
+            return Ok(());
+        }
+        self.set_clock_status(enabled, RadioPeripherals::Phy);
+        Ok(())
+    }
+    /// Enable or disable a modem clock.
+    ///
+    /// If required, this will also enabled or disable the PHY clock.
+    /// # Panics
+    /// This panics, if [RadioPeripherals::Phy] is passed in, since the PHY
+    /// clock is controlled internally.
+    pub fn set_modem_clock_status(&self, enabled: bool, radio_peripheral: RadioPeripherals) {
+        // We first assert, that the clock we're touching isn't the PHY clock, since
+        // that's special.
+        assert_ne!(radio_peripheral, RadioPeripherals::Phy);
+
+        // We obtain a reference to the AtomicBool tracking the state of the clock.
+        let modem_clock_enabled = match radio_peripheral {
+            #[cfg(bt)]
+            RadioPeripherals::Bt => &self.bt_clock_enabled,
+            #[cfg(wifi)]
+            RadioPeripherals::Wifi => &self.wifi_clock_enabled,
+            #[cfg(ieee802154)]
+            RadioPeripherals::Ieee802154 => &self.ieee802154_clock_enabled,
+            _ => unreachable!(),
+        };
+        // If the states already match, we return. Otherwise the new state is stored.
+        if modem_clock_enabled
+            .compare_exchange(!enabled, !enabled, Ordering::Relaxed, Ordering::Relaxed)
+            .is_err()
+        {
+            return;
+        }
+
+        if enabled {
+            // If the PHY clock isn't enabled yet, we disable it.
+            let _ = self.set_phy_clock_status(true);
+            self.set_clock_status(true, radio_peripheral);
+        } else {
+            self.set_clock_status(false, radio_peripheral);
+            let _ = self.set_phy_clock_status(false);
+        }
```

Comment:
I think this whole condition should be inside `set_clock_status`

Reply:
I'm open to moving it there, but I initially didn't do this, since we already call `set_clock_status` in `set_phy_clock_status` and we would then have to put a lot more checks into `set_clock_status`.

Reply:
My reasoning is that it looks unlikely that we want to switch PHY independently from the other clocks, so I'd hide the complexity behind a single call that does everything it needs to. I leave it up to you, I guess this doesn't really relate to the design here.

Reply:
That's a good point. Theoretically, we could also just remove the `RadioPeripherals::Phy` variant and handle that through a separate function, so we wouldn't have to panic if the user passes it in.

Reply:
I have moved the clock setting to another functioning handling this.